python/plugins/db_manager/db_plugins/vlayers/data_model.py

```diff
@@ -63,40 +70,86 @@ def rowCount(self, index=None):
         return 0
 
 
-class LSqlResultModel(BaseTableModel):
-    # BaseTableModel
+class LSqlResultModelTask(SqlResultModelTask):
 
-    def __init__(self, db, sql, parent=None):
-        # create a virtual layer with non-geometry results
-        t = QTime()
-        t.start()
+    def __init__(self, db, sql, parent):
+        SqlResultModelTask.__init__(self, db, sql, parent)
 
         tf = QTemporaryFile()
         tf.open()
-        tmp = tf.fileName()
+        path = tf.fileName()
         tf.close()
 
         df = QgsVirtualLayerDefinition()
-        df.setFilePath(tmp)
+        df.setFilePath(path)
         df.setQuery(sql)
-        p = QgsVectorLayer(df.toString(), "vv", "virtual")
-        self._secs = t.elapsed() / 1000.0
 
-        if not p.isValid():
-            data = []
-            header = []
-            raise DbError(p.dataProvider().error().summary(), sql)
+        self.subtask = QgsVirtualLayerTask(df)
+        self.addSubTask(self.subtask, [], QgsTask.ParentDependsOnSubTask)
+
+    def run(self):
+        try:
+            path = self.subtask.definition().filePath()
+            sql = self.subtask.definition().query()
+            self.model = LSqlResultModel(self.db, sql, None, self.subtask.layer(), path)
+        except Exception as e:
+            self.error = BaseError(str(e))
+            return False
+        return True
+
+    def cancel(self):
+        self.subtask.cancel()
```

Comment:
Are you sure this is needed? The task manager should automatically take care of this.

Reply:
OK good to know. I removed it so.